### PR TITLE
Tests: Add Tests to `contains` Filter Operator on PostgreSQL Backend

### PR DESCRIPTION
**Pull request by rabbull**
This PR adds tests to validate the behavior of the `contains` operator when applied to JSON objects and arrays in the PostgreSQL backend. The tests cover a variety of scenarios, including type matching, negations, and edge cases such as empty arrays or objects. This is a test-only PR, and no features or functionality in aiida-core have been modified.

Some corner case behaviors appear unintuitive and will be documented in a separate issue for further investigation.

### CODE REVIEW

**Review by GeigerJ2 — APPROVED** [review]
Great work, @rabbull, thanks for adding those extensive tests. I'd say we get this merged, and then figure out how we can resolve the unexpected behaviors.

**Comment by rabbull** [comment]
> I'd use the first paragraph of this PR description as the commit message for the squashed commit.

Sure, this LGTM. :)

File: tests/orm/test_querybuilder.py [hunk 1]
```diff
@@ -1702,3 +1703,159 @@ def test_statistics_default_class(self, aiida_localhost):
...
+                        'k2': '2',
+                        'k3': None,
+                    }
+                },
+                {'attributes.dict': {'!contains': {}}},
+                False,
+            ),
+            # TODO: these pass, but why? are these behaviors expected?
+            # non-exist `attr_key`s
+            ({'map': {}}, {'attributes.dict': {'contains': {}}}, False),
```
Comment (GeigerJ2) [inline tests/orm/test_querybuilder.py:1848]: Probably same reason as above.

File: tests/orm/test_querybuilder.py [hunk 2]
```diff
@@ -1702,3 +1703,159 @@ def test_statistics_default_class(self, aiida_localhost):
...
+        orm.Dict(data).store()
+        qb = orm.QueryBuilder().append(orm.Dict, filters=filters)
+        assert qb.count() in {0, 1}
+        found = qb.count() == 1
+        assert found == is_match
+
+    @pytest.mark.parametrize(
+        'data,filters,is_match',
+        (
+            # contains different types of values
```
Comment (GeigerJ2) [inline tests/orm/test_querybuilder.py:1754]: Could you also add some tests on the keys, please.

**Review by GeigerJ2 — CHANGES_REQUESTED** [review]
Very nice work, thanks a lot for extending the tests so thoroughly, @rabbull! Some minor proposals to extend them further.

File: tests/orm/test_querybuilder.py [hunk 3]
```diff
@@ -1702,3 +1703,159 @@ def test_statistics_default_class(self, aiida_localhost):
...
+            ({'arr': [1, '2', None]}, {'attributes.arr': {'contains': [1]}}, True),
+            ({'arr': [1, '2', None]}, {'attributes.arr': {'contains': ['2']}}, True),
+            ({'arr': [1, '2', None]}, {'attributes.arr': {'contains': [None]}}, True),
+            # contains multiple elements of various types
+            ({'arr': [1, '2', None]}, {'attributes.arr': {'contains': [1, None]}}, True),
+            # contains non-exist elements
+            ({'arr': [1, '2', None]}, {'attributes.arr': {'contains': [114514]}}, False),
+            # contains empty set
+            ({'arr': [1, '2', None]}, {'attributes.arr': {'contains': []}}, True),
+            ({'arr': []}, {'attributes.arr': {'contains': []}}, True),
```
[the remark this answers is not in the document]
Comment (rabbull) [reply tests/orm/test_querybuilder.py:1722]: It's already added at L1731

File: tests/orm/test_querybuilder.py (hunk 2 above)
Comment (rabbull) [reply tests/orm/test_querybuilder.py:1754]: Explained in a previous comment.

File: tests/orm/test_querybuilder.py [hunk 4]
```diff
@@ -1702,3 +1703,159 @@ def test_statistics_default_class(self, aiida_localhost):
...
+            # ({'arr': [[2, 3], [1]]}, {'attributes.arr': {'contains': [[2]]}}, False),
+            # negations
+            ({'arr': [1, '2', None]}, {'attributes.arr': {'!contains': [1]}}, False),
+            ({'arr': [1, '2', None]}, {'attributes.arr': {'!contains': []}}, False),
+            ({'arr': [1, '2', None]}, {'attributes.arr': {'!contains': [114514]}}, True),
+            ({'arr': [1, '2', None]}, {'attributes.arr': {'!contains': [1, 114514]}}, True),
+            # TODO: these pass, but why? are these behaviors expected?
+            # non-exist `attr_key`s
+            ({'foo': []}, {'attributes.arr': {'contains': []}}, False),
+            ({'foo': []}, {'attributes.arr': {'!contains': []}}, False),
```
[the remark this answers is not in the document]
Comment (rabbull) [reply tests/orm/test_querybuilder.py:1737]: Ach so, this explains! I would suggest we leave this behavior undefined. In another word, if a user constructs a query on non-existent keys, AiiDA shall return arbitrary results, and in this specific case, the return will always be `False`.

That said, regarding the extended test cases, the `contains` operator does not support checking the existence of keys within JSON objects. If the intention is to verify the presence of specific keys, the `has_key` operator should be used instead. This distinction also explains why no "key" tests are included in the tests for JSON objects.

File: tests/orm/test_querybuilder.py [hunk 5]
```diff
@@ -1702,3 +1703,159 @@ def test_statistics_default_class(self, aiida_localhost):
...
+            ({'arr': [1, '2', None]}, {'attributes.arr': {'contains': [None]}}, True),
+            # contains multiple elements of various types
+            ({'arr': [1, '2', None]}, {'attributes.arr': {'contains': [1, None]}}, True),
+            # contains non-exist elements
+            ({'arr': [1, '2', None]}, {'attributes.arr': {'contains': [114514]}}, False),
+            # contains empty set
+            ({'arr': [1, '2', None]}, {'attributes.arr': {'contains': []}}, True),
+            ({'arr': []}, {'attributes.arr': {'contains': []}}, True),
+            # nested arrays
+            ({'arr': [[1, 0], [0, 2]]}, {'attributes.arr': {'contains': [[1, 0]]}}, True),
```
[the remark this answers is not in the document]
Comment (rabbull) [reply tests/orm/test_querybuilder.py:1724]: These tests are added, but I made a little modification to make them work as intended:
```python
            ({'arr': [[1, 0], [0, 2]]}, {'attributes.arr': {'contains': [[3]]}}, False),
            ({'arr': [[1, 0], [0, 2]]}, {'attributes.arr': {'contains': [3]}}, False),
            ({'arr': [[1, 0], [0, 2]]}, {'attributes.arr': {'contains': [[2]]}}, True),
            ({'arr': [[1, 0], [0, 2]]}, {'attributes.arr': {'contains': [2]}}, False),
            ({'arr': [[1, 0], [0, 2], 3]}, {'attributes.arr': {'contains': [[3]]}}, False),
            ({'arr': [[1, 0], [0, 2], 3]}, {'attributes.arr': {'contains': [3]}}, True),
```

[the remark this answers is not in the document]
Comment (GeigerJ2) [reply tests/orm/test_querybuilder.py:1724]: Looks good! Interesting that:

```python
            ({'arr': [[1, 0], [0, 2], 3]}, {'attributes.arr': {'contains': 3}}, True),
```

gives:

```
>           raise ex.with_traceback(None)
E           psycopg.errors.UndefinedFunction: operator does not exist: jsonb @> integer
E           LINE 4: ...ND (CAST((db_dbnode_1.attributes #> $2) AS JSONB) @> $3::INT...
E                                                                        ^
E           HINT:  No operator matches the given name and argument types. You might need to add explicit type casts.
```

And same behavior for:

```python
            ({'foo': []}, {'attributes': {'!contains': 'foo'}}, True),
```

Just putting this here for future reference that it's actually an invalid query due to the type mismatch.

File: tests/orm/test_querybuilder.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (rabbull) [reply tests/orm/test_querybuilder.py:1741]: > Just to confirm: So you are also OK with the behavior?

Yes, this looks reasonable to me.

**Review by GeigerJ2 — APPROVED** [review]
